osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -171,7 +176,22 @@ public void SelectRandom()
             if (visibleGroups.Count < 1)
                 return;
 
-            BeatmapGroup group = visibleGroups[RNG.Next(visibleGroups.Count)];
+            BeatmapGroup group;
+            if (randomType == SelectionRandomType.RandomPermutation)
+            {
+                List<BeatmapGroup> notSeenGroups = visibleGroups.Except(seenGroups).ToList();

```

Comment:
This can become very inefficient with large sets of beatmaps. How about using Knuth shuffle (https://stackoverflow.com/questions/14535274/randomly-permutation-of-n-consecutive-integer-number ) on a the integers 1 to `groups.Count` whenever the amount of groups changes, and simply picking the next index in the shuffled list whenever F2 is pressed. That also has the benefit of not needing to store beatmap groups in an additional data structure (your hash set).